### DISCUSSION TITLE
Does react-query act as a state management library? Can I use it with Redux?

### DISCUSSION THREAD

**Discussion by ghost** [TanStack/query#329] [General] [answered]
does react-query act as a state management library?
 Can I intergrate it with redux or easy-peasy?

Pls someone give an example or some repo or tutorial post

Regards,
A React newbie

**Comment by tannerlinsley** [comment, marked as answer]
React Query is a *type* of state manager, specifically designed to manage asynchronous state that you get from *outside* of your app, so server-state, API state, or anything else that is not local client state.  With that said, it's easy and even encouraged to keep use React Query along side a global state manager for your *client* state.

Its very common for existing applications that after moving their *server* state to React Query, their *client* state is extremely small and doesn't even need an external library like Redux. However, some applications actually do have a lot of local client state to manage and something like Redux is warranted. 

I will link to a talk about this very subject very soon. But you can feel good pressing forward using both React Query for anything that is asynchronous data and Redux for anything that is local and synchronous.

**Reply by Miteshdv** [reply]
> React Query is a _type_ of state manager, specifically designed to manage asynchronous state that you get from _outside_ of your app, so server-state, API state, or anything else that is not local client state. With that said, it's easy and even encouraged to keep use React Query along side a global state manager for your _client_ state.
> 
> Its very common for existing applications that after moving their _server_ state to React Query, their _client_ state is extremely small and doesn't even need an external library like Redux. However, some applications actually do have a lot of local client state to manage and something like Redux is warranted.
> 
> I will link to a talk about this very subject very soon. But you can feel good pressing forward using both React Query for anything that is asynchronous data and Redux for anything that is local and synchronous.

Hi Tanner

 React Query is awesome .Please help me  to undersatnd ,How we can get the Control on Asynchronous calls that Redux saga gives ..  Another thing when using redux ,component has only component logic and the API calls reside in Thunks/ Sagas .. But what i see is more lines of code added to component What are your thoughts on this

**Comment by aashish254** [comment]
**# Complete Guide: React Query + Redux Integration Patterns**\n\nGreat question! Yes, you absolutely CAN use React Query alongside Redux (or any other state manager). Here's a comprehensive guide with working examples.\n\n## Understanding the Separation\n\n**React Query**: Manages **server state** (async data from APIs)  \n**Redux**: Manages **client state** (UI state, local preferences, form data, etc.)\n\n## Pattern 1: Minimal Redux + Full React Query (Recommended)\n\nThis is the most common pattern - use minimal Redux only for what you truly need locally:\n\n```typescript\n// redux/store.ts\nimport { configureStore } from '@reduxjs/toolkit';\n\nexport const store = configureStore({\n  reducer: {\n    // Only client-side state needed here\n    ui: uiReducer,        // Theme, sidebar expanded/collapsed\n    userSettings: settingsReducer, // User preferences, dark mode toggle\n    cart: cartReducer,    // Shopping cart before checkout\n  },\n});\n\nexport type RootState = ReturnType<typeof store.getState>;\nexport type AppDispatch = typeof store.dispatch;\n```\n\n```typescript\n// components/UserProfile.tsx - Uses React Query\nimport { useQuery } from '@tanstack/react-query';\n\nfunction UserProfile({ userId }: { userId: string }) {\n  // Server state managed by React Query\n  const { data: user, isLoading, error } = useQuery({\n    queryKey: ['user', userId],\n    queryFn: () => fetchUser(userId),\n  });\n\n  return (\n    <div>\n      {isLoading && <p>Loading...</p>}\n      {error && <p>Error loading user</p>}\n      {user && (\n        <>\n          <h1>{user.name}</h1>\n          <p>Email: {user.email}</p>\n        </>\n      )}\n    </div>\n  );\n}\n```\n\n```typescript\n// components/ThemeProvider.tsx - Uses Redux\nimport { useSelector } from 'react-redux';\n\nfunction ThemeProvider() {\n  // Client state managed by Redux\n  const theme = useSelector((state: RootState) => state.ui.theme);\n\n  return (\n    <div className={`theme-${theme}`}>\n      {/* Your app content */}\n    </div>\n  );\n}\n```\n\n## Pattern 2: Full-Featured Example\n\nHere's a real-world e-commerce setup:\n\n```typescript\n// api/users.ts - React Query hook for server data\nexport function useUserProfile(userId: string) {\n  return useQuery({\n    queryKey: ['users', userId],\n    queryFn: async () => {\n      const res = await fetch(`/api/users/${userId}`);\n      if (!res.ok) throw new Error('Failed to fetch user');\n      return res.json();\n    },\n    staleTime: 5 * 60 * 1000, // 5 minutes\n  });\n}\n\nexport function useUpdateUser() {\n  const queryClient = useQueryClient();\n  \n  return useMutation({\n    mutationFn: updateUserApi,\n    onSuccess: () => {\n      // Invalidate and refetch user list\n      queryClient.invalidateQueries({ queryKey: ['users'] });\n    },\n  });\n}\n\n// redux/slices/cartSlice.ts - Redux for cart state\nexport const cartSlice = createSlice({\n  name: 'cart',\n  initialState: {\n    items: [] as CartItem[],\n    isOpen: false,\n  },\n  reducers: {\n    addItem: (state, action: PayloadAction<CartItem>) => {\n      state.items.push(action.payload);\n    },\n    removeItem: (state, action: PayloadAction<string>) => {\n      state.items = state.items.filter(item => item.id !== action.payload);\n    },\n    toggleCart: (state) => {\n      state.isOpen = !state.isOpen;\n    },\n  },\n});\n\nexport const { addItem, removeItem, toggleCart } = cartSlice.actions;\n```\n\n```typescript\n// pages/product/[id].tsx - Combining both\nimport { useQuery, useMutation } from '@tanstack/react-query';\nimport { useDispatch, useSelector } from 'react-redux';\nimport { toggleCart, addItem } from '@/redux/slices/cartSlice';\n\nfunction ProductPage({ productId }: { productId: string }) {\n  // Server state via React Query\n  const { data: product, isLoading } = useProduct(productId);\n  \n  // Client state via Redux\n  const dispatch = useDispatch();\n  const cartItems = useSelector((state: RootState) => state.cart.items);\n  \n  // Mutation via React Query\n  const addToCart = useMutation({\n    mutationFn: (item: Product) => addProductToCart(item),\n    onSuccess: (_, product) => {\n      dispatch(addItem(product));\n      dispatch(toggleCart()); // Open cart drawer\n    },\n  });\n\n  if (isLoading) return <LoadingSpinner />;\n\n  return (\n    <div>\n      <h1>{product?.name}</h1>\n      <button onClick={() => addToCart.mutate(product)}>\n        Add to Cart ({cartItems.length})\n      </button>\n    </div>\n  );\n}\n```\n\n## Pattern 3: When You Might Need Both Heavily\n\nSome apps genuinely need significant amounts of both:\n\n```typescript\n// Complex dashboard with lots of UI state AND server data\n\n// Redux for complex client interactions\nconst dashboardSlice = createSlice({\n  name: 'dashboard',\n  reducers: {\n    setWidgetPosition: (state, action) => {\n      // Drag-and-drop widget positions\n      state.widgets[action.payload.id] = action.payload.position;\n    },\n    filterData: (state, action) => {\n      // Local filters applied before sending to API\n      state.filters = action.payload;\n    },\n    sortData: (state, action) => {\n      // Local sorting preferences\n      state.sortConfig = action.payload;\n    },\n  },\n});\n\n// React Query for all API calls\nconst useDashboardData = (filters: DashboardFilters) => {\n  return useQuery({\n    queryKey: ['dashboard-data', filters],\n    queryFn: () => fetchDashboardData(filters),\n    select: (data) => applyLocalSorting(data, state.sortConfig),\n  });\n};\n```\n\n## Integration Examples\n\n### Example A: Prefill Form from Redux then Submit via React Query\n\n```typescript\nfunction CreateUserForm() {\n  const initialUser = useSelector(selectInitialUserData);\n  const [formData, setFormData] = useState(initialUser || {});\n  \n  const submitMutation = useMutation({\n    mutationFn: createUser,\n    onSuccess: () => {\n      // After successful API call, update Redux cache\n      queryClient.setQueryData(['currentUser'], newUser);\n    },\n  });\n\n  return (\n    <form onSubmit={handleSubmit(submitMutation)}>\n      <input\n        value={formData.name}\n        onChange={(e) => setFormData({...formData, name: e.target.value})}\n      />\n    </form>\n  );\n}\n```\n\n### Example B: Real-time Updates\n\n```typescript\n// Subscribe to Redux changes and refetch React Query\nfunction AutoRefetcher() {\n  const dispatch = useDispatch();\n  const queryClient = useQueryClient();\n\n  useEffect(() => {\n    const unsubscribe = store.subscribe(() => {\n      const state = store.getState();\n      \n      // If user settings changed, refetch relevant queries\n      if (state.userSettings.dataFetched) {\n        queryClient.refetchQueries(['userData']);\n      }\n    });\n\n    return () => unsubscribe();\n  }, [queryClient]);\n\n  return null; // This is just a side-effect component\n}\n```\n\n## Best Practices Summary\n\n✅ **DO**: Use React Query for all server/state async data  \n✅ **DO**: Use Redux for UI/local state that doesn't come from API  \n✅ **DO**: Keep Redux slice logic simple, delegate complex logic to React Query mutations  \n❌ **DON'T**: Store server data in Redux (let React Query manage it)  \n❌ **DON'T**: Try to \"force\" React Query into Redux (it adds unnecessary complexity)  \n\n## When to Choose Each\n\n| State Type | Tool | Why |\n|------------|------|-----|\n| API responses, pagination, caching | React Query | Built-in optimizes, background refresh, retries |\n| Form state, modals, dropdowns | Redux (or Context) | Simple subscription-based updates |\n| Theme preferences | Redux | Global sync state across app |\n| Shopping cart | Redux | Immediate local updates needed |\n| User profile data | React Query | Needs caching, validation, stale data handling |\n| Notification system | Redux or separate | Event-driven state management |\n\n## Complete Working Setup\n\n```typescript\n// _app.tsx\nfunction MyApp({ Component, pageProps }) {\n  const [queryClient] = useState(() => new QueryClient({\n    defaultOptions: {\n      queries: {\n        staleTime: 5 * 60 * 1000, // 5 min\n      },\n    },\n  }));\n\n  return (\n    <ReduxProvider store={store}>\n      <QueryClientProvider client={queryClient}>\n        <ThemeProvider>\n          <AutoRefetcher />\n          <Component {...pageProps} />\n        </ThemeProvider>\n      </QueryClientProvider>\n    </ReduxProvider>\n  );\n}\n```\n\n---\n\n**TL;DR**: Yes, combine them! React Query for server state (APIs), Redux for client state (UI, local prefs). They complement each other perfectly. Start with this approach and adjust based on your specific needs. 🚀
